File: src/leetCode/23.实现strstr().cpp
class Solution {
public:
    int strStr(string haystack, string needle) {
        if(needle.empty()){
            return 0;
        }
        vector<int>nextVal = getNextVal(needle);
        int strLen = haystack.length(),i = 0;
        int subLen = needle.length(),j = 0;
        while(i<strLen && j<subLen){
            if( j == -1 || haystack[i] == needle[j]){
                i++;
                j++;
            }
            else{
                j = nextVal[j];
            }
        }
        if(j ==  subLen){
            return i - j;
        }
        else{
            return -1;
        }
    }
   
   vector<int>getNextVal(string substr) {
	int len = substr.length();
	vector<int>nextVal(len+1, -1);
	int i = 0,j = -1;
	while (i < len) {
		if (j == -1 || substr[i] == substr[j]) {
			i++;
			j++;
			if (substr[i] != substr[j]) {
				nextVal[i] = j;
			}
			else {
				nextVal[i] = nextVal[j];
			}
		}
		else {
			j = nextVal[j];
		}
	}
	return nextVal;
    }
};

File: src/leetCode/23.实现strStr().js
/**
 * @param {string} haystack
 * @param {string} needle
 * @return {number}
 */
var strStr = function (haystack, needle) {
  if (needle === "") {
    return 0;
  }
  let i = 0,
    m = haystack.length;
  let j = 0,
    n = needle.length;
  let next = getNext(needle);
  while (i < m && j < n) {
    if (j == -1 || needle[j] === haystack[i]) {
      i++;
      j++;
    } else {
      j = next[j];
    }
  }
  return j === n ? i - j : -1;
};

function getNext(needle) {
  let next = [];
  next[0] = -1;
  let k = -1,
    j = 0;
  while (j < needle.length) {
    if (k == -1 || needle[k] === needle[j]) {
      next[++j] = ++k;
    } else {
      k = next[k]; //后缀指针k回退到next[k]的位置
    }
  }
  return next;
}
